### ISSUE TITLE
Render images as MathJax

### ISSUE THREAD

**Issue by thenoncoder** [mathjax/MathJax#1266]
My scripting skills aren't that good, so I'm looking for someone to kindly provide a userscript for this.

I've noticed that websites have latex in the alt text of their maths images, so I thought that it could be possible for someone to make a userscript that converts that latex to MathJax.

I *really* hate those blurry maths png images, that's why!

The advantage of such a userscript is that it would work on virtually almost *all* sorts of websites, instead of having to make a special userscript for every single website.

The only thing though is that if the alt text turns out to be *not* latex, then the unserscript should be able to understand that, and not break the page trying to convert an ordinary image into MathJax.

Thanks a lot! :)

**Comment by pkra** [comment]
Here's an [old bookmarklet I once created for Wikipedia images](https://gist.github.com/pkra/5500316). It should give you  a good point to start. The trouble with Wikipedia (and other sites) are custom macros. Keep in mind that some sites like mathworld use ascii-like syntax (though not actual asciimath, it usually renders ok).

**Comment by thenoncoder** [comment]
Thanks!

It works, but the text is really small.

On actual MathJax text you could click on it to zoom, but apparently your script doesn't have the right click menu.

But it's good that it just works, as I've tried too many scripts and none of them did.

Thanks again! :)

**Comment by thenoncoder** [comment]
Is it possible to make the default zoom 200%?

I hate having to set it for every website.

**Comment by dpvc** [comment]
There is also a [Greasemonkey script](http://docs.mathjax.org/en/latest/dynamic.html#mathjax-and-greasemonkey) that does the same sort of thing, but that doesn't use jQuery (I worry about inserting jQuery into an arbitrary page, as it might interfere with the page's own copy of jQuery, especially if it is a different version of jQuery).

You could have the script add configuration that sets the zoom scaling to 200.  Adding `MathJax.Hub.Config({menuSettings:zoom:200})` would do it, I think.

**Comment by thenoncoder** [comment]
I actually tried that script before and it didn't work.

**Comment by thenoncoder** [comment]
Most websites have latex in the image alt, not as a`<math>` element, and the latter is what the userscript is trying to do.

**Comment by thenoncoder** [comment]
I don't know how to make a proper userscript to convert image alts to latex.

It'd have to ignore non-latex image alts, and again, I have no idea how to do that.

**Comment by pkra** [comment]
> It'd have to ignore non-latex image alts, and again, I have no idea how to do that.

Yes, that's a problem you can only work around with additional information (such as a specific class in the case of Wikipedia).

**Comment by dpvc** [comment]
> Most websites have latex in the image alt, not as a `<math>` element, and the latter is what the userscript is trying to do.

I think you may be looking at the wrong user script.  There are two in that section, and it is the *second* one.  That one looks for image tags and takes the TeX out of the `alt` attribute.

**Comment by thenoncoder** [comment]
@dpvc I didn't look at the wrong one, I'm not that stupid bro.

I looked at specifically, this one : 

`// ==UserScript==
// @name           MathJax in Wikipedia
// @namespace      http://www.mathjax.org/
// @description    Insert MathJax into Wikipedia pages
// @include        http://en.wikipedia.org/wiki/*
// ==/UserScript==

if ((window.unsafeWindow == null ? window : unsafeWindow).MathJax == null) {
  //
  //  Replace the images with MathJax scripts of type math/tex
  //
  var images = document.getElementsByTagName('img'), count = 0;
  for (var i = images.length - 1; i >= 0; i--) {
    var img = images[i];
    if (img.className === "tex") {
      var script = document.createElement("script"); script.type = "math/tex";
      if (window.opera) {script.innerHTML = img.alt} else {script.text = img.alt}
      img.parentNode.replaceChild(script,img); count++;
    }
  }
  if (count) {
    //
    //  Load MathJax and have it process the page
    //
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML-full";
    document.getElementsByTagName("head")[0].appendChild(script);
  }`

And it didn't work.

**Comment by thenoncoder** [comment]
@pkra But most websites don't have a specific class. They just have an ordinary image alt.

If a userscript was made to convert that image alt to MathJax, then it would work on *plenty* of websites.

I just absolutely hate those pixelated, blurry, png images.

**Comment by dpvc** [comment]
> I didn't look at the wrong one, I'm not that stupid bro.

Since you said the user script was trying to "do" `<math>` elements, I didn't see how you were looking at the right one, since the second one makes no mention of math elements in any way.  It is looking at `<img>` tags and inserting the `<script>` tags that MathJax uses.  There are no `<math>` tags involved. 

The script tags are set to contain the LaTeX from the image `alt` attributes and the MIME type is set so that MathJax will process them as TeX.  That sounds very close to what you are trying to do.

It is true that the user script is designed for use only on Wikipedia, so it looks for the images with `className` of "tex", which is how they were marked at the time the script was written (I haven't checked lately).  The `@include` directive also is set so that the user script only runs on the English Wikipedia site.  So you certainly will have to make modifications.  It was only intended as a starting point, not a finished solution for you.

Also, "it didn't work" is not a very specific comment.  You have not indicated what you tried, or in what way it failed, or what types of error message you may have gotten, or any real information at all.  So I'm not sure what you expect from us.

> If a userscript was made to convert that image alt to MathJax, then it would work on plenty of websites.

If the `alt` attribute is in (La)TeX form, then the script you link to should be easily modified to do what you ask.  It already inserts the DOM elements needed by MathJax.  It is only the `className` test that would need to be removed.  Of course, if you are asking for a way to distinguish LaTeX `alt` attributes from other `alt` text, I do not know of a reasonable way to do that.  If you have a means of doing that, by all means, go ahead and add that to the script.

**Comment by Herst** [comment]
For you and anyone else who might be interested in such bookmarklets, have a look at: http://www.informatik-forum.at/showthread.php?t=107276 (site is German but the code should be pretty self-explanatory)

**Comment by thenoncoder** [comment]
@Herst, thanks.

It doesn't work on any website, though, :(

**Comment by Herst** [comment]
Of course not, this one is very specific, I meant that you can use it as a basis for your own bookmarklet.

**Comment by thenoncoder** [comment]
@herst Okay, then. I'll see what I can do with it.
